### PR TITLE
Bootstrap test refactor

### PR DESCRIPTION
**Pull request by ycliuhw**
Refactor k8s bootstrapSuite to use fake client rather gomock;

### CODE REVIEW

**Comment by ycliuhw** [comment]
[refactor CAAS bootstrap tests to use fake client rather than gomock](https://trello.com/c/2RGXtdNx/1769-refactor-caas-bootstrap-tests-to-use-fake-client-rather-than-gomock)

**Comment by ycliuhw** [comment]
[refactor CAAS bootstrap tests to use fake client rather than gomock](https://trello.com/c/Cc3FmC0i/296-refactor-caas-bootstrap-tests-to-use-fake-client-rather-than-gomock)

File: caas/kubernetes/provider/bootstrap_test.go [hunk 1]
```diff
@@ -812,246 +788,140 @@ JUJU_DEV_FEATURE_FLAGS=developer-mode $JUJU_TOOLS_DIR/jujud machine --data-dir $
...
+		close(done)
+	})
 	go func() {
 		errChan <- controllerStacker.Deploy()
 	}()
-
-	err = s.clock.WaitAdvance(3*time.Second, testing.LongWait, 1)
-	c.Assert(err, jc.ErrorIsNil)
+	go func() {
+		svcDone := time.After(0 * time.Second)
```
Comment (hpidcock) [inline caas/kubernetes/provider/bootstrap_test.go:800]: Can we use a s.mockServices.Watch?

File: caas/kubernetes/provider/bootstrap_test.go [hunk 2]
```diff
@@ -812,246 +788,140 @@ JUJU_DEV_FEATURE_FLAGS=developer-mode $JUJU_TOOLS_DIR/jujud machine --data-dir $
...
+		podDone := time.After(0 * time.Second)
```
Comment (hpidcock) [inline caas/kubernetes/provider/bootstrap_test.go:801]: s.mockStatefulset.Watch?

**Review by hpidcock — APPROVED** [review]
LGTM just an issue with how we are waiting for resources

File: caas/kubernetes/provider/bootstrap_test.go (hunk 1 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/bootstrap_test.go:800]: Unfortunately, I don't think those event side effects are triggered automatically by the service gets created/updated... etc.
Just like the pods are not created even we created the Statefulset.

Comment (ycliuhw) [reply caas/kubernetes/provider/bootstrap_test.go:800]: Ah, my previous testing was wrong.
Changed to use the fake watcher now.

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**